pkg/agent/nodeportlocal/portcache/port_table_windows.go

```diff
@@ -0,0 +1,283 @@
+			return 0, err
+		}
+
+		protocolSocketData.State = stateInUse
+	}
+	return npData.NodePort, nil
+}
+
+// syncRules ensures that contents of the port table matches the iptables rules present on the Node.
+func (pt *PortTable) syncRules() error {
+	pt.tableLock.Lock()
+	defer pt.tableLock.Unlock()
+	nplPorts := make([]rules.PodNodePort, 0, len(pt.NodePortTable))
+	for _, npData := range pt.NodePortTable {
+		protocols := make([]string, 0, len(supportedProtocols))
+		for _, protocol := range npData.Protocols {
+			if protocol.State == stateInUse {
+				protocols = append(protocols, protocol.Protocol)
+			}
+		}
+		nplPorts = append(nplPorts, rules.PodNodePort{
+			NodePort:  npData.NodePort,
+			PodPort:   npData.PodPort,
+			PodIP:     npData.PodIP,
+			Protocols: protocols,
+		})
+	}
+	if err := pt.PodPortRules.AddAllRules(nplPorts); err != nil {
+		return err
+	}
+	return nil
+}
+
+// RestoreRules should be called on startup to restore a set of NPL rules. It is non-blocking but
+// takes as a parameter a channel, synced, which will be closed when the necessary rules have been
+// restored successfully. No other operations should be performed on the PortTable until the channel
+// is closed.
+func (pt *PortTable) RestoreRules(allNPLPorts []rules.PodNodePort, synced chan<- struct{}) error {
+	pt.tableLock.Lock()
+	defer pt.tableLock.Unlock()
+	for _, nplPort := range allNPLPorts {
+		protocols, err := addruleForPort(pt.PodPortRules, nplPort.NodePort, nplPort.PodIP, nplPort.PodPort, nplPort.Protocols[0])
+		if err != nil {
+			// This will be handled gracefully by the NPL controller: if there is an
+			// annotation using this port, it will be removed and replaced with a new
+			// one with a valid port mapping.
+			klog.ErrorS(err, "Cannot bind to local port, skipping it", "port", nplPort.NodePort)
+			updateCloseState(protocols)
+			continue
+		}
+
+		npData := &NodePortData{
+			NodePort:  nplPort.NodePort,
+			PodPort:   nplPort.PodPort,
+			PodIP:     nplPort.PodIP,
+			Protocols: protocols,
+		}
+		for _, protocol := range nplPort.Protocols {
+			protocolSocketData := npData.FindProtocol(protocol)
+			if protocolSocketData == nil {
+				return fmt.Errorf("unknown protocol %s", protocol)
+			}
+			protocolSocketData.State = stateInUse
+		}
+		pt.NodePortTable[nplPort.NodePort] = npData
+		pt.PodEndpointTable[podIPPortFormat(nplPort.PodIP, nplPort.PodPort)] = pt.NodePortTable[nplPort.NodePort]
+	}
+	// retry mechanism as iptables-restore can fail if other components (in Antrea or other
+	// software) are accessing iptables.
+	go func() {
+		defer close(synced)
+		var backoffTime = 2 * time.Second
+		for {
+			if err := pt.syncRules(); err != nil {
+				klog.ErrorS(err, "Failed to restore netnat rules", "backoff", backoffTime)
+				time.Sleep(backoffTime)
+				continue
+			}
+			break
+		}
+	}()
+	return nil
+}
+
+func (d *NodePortData) UpdateCloseStates() error {
```

Comment:
Do we need to call it outside the current package?

Reply:
Updated.